### ISSUE TITLE
support non-Bluesky ATProto apps, eg Whitewind

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#1178]
https://whtwnd.com/ is an early third party ATProto service with its own lexicons (for long form writing) and maybe appview? We should test interop!
I tried liking a post, https://snarfed.org/2024-07-11_notes-on-running-a-full-network-atproto-relay-july-2024-bryan-newbold-whitewind-blog =>
https://whtwnd.com/bnewbold.net/3kwzl7tye6u2y , but it didn't work because we don't translate whtwnd.com URLs yet.

**Comment by Tamschi** [comment]
It has an own (private) AppView I believe, yes.

I'm not sure in how far ActivityPub-`Article`s could be bridged into WhiteWind since the latter uses Markdown as storage format (maybe just `Article`s that are natively Markdown on ActivityPub?), but WhiteWind posts could definitely be rendered as AP-`Article` for links in Mastodon and for display in other software that has more direct support for the format.

Note that formally WhiteWind isn't a "Bluesky" app (it's actually largely incompatible with that, as far as its own data goes) but sits independently on ATProto. I'll still label this `app`, though.

**Comment by snarfed** [comment]
Yes! And re Markdown, when Bridgy Fed converts AP HTML content to plain text for ATProto, it's actually Markdown, just a somewhat minimal form, so we already have that more or less for free.

**Comment by snarfed** [comment]
Expanding this to cover all non-Bluesky (ie non-`app.bsky.*`) ATProto apps. There are lots! UX is maybe the biggest question. Do we want to integrate with them all automatically? Ie when a user bridges an account to/from Bluesky, are they bridging just the app? Or the whole protocol and all aps in the ecosystem?

Concrete examples: for a bridged web site, would we publish its articles to WhiteWind as well as Bluesky? If there's an ATProto polls app, and someone posts a poll on Mastodon, would we bridge it into that ATProto app?

Mechanically, for each app, we'd need to add granary support for converting its lexicons to/from AS1, and then Bridgy Fed support for its semantics. Straightforward, but only after we decide on the UX behavior.

**Comment by snarfed** [comment]
Coming back to this now, I think this is simpler than I worried earlier. For fedi => Bluesky, we can implement whichever lexicons we want, even multiple, eg we could convert AS2 events to both Smoke Signal and https://lexicon.community/ lexicons. For Bluesky => fedi, we can bridge whichever lexicons the user publishes.

Maybe the only question is when there are multiple lexicons for core types like profiles, or follows, and a Bluesky user publishes more than one of them. Which one do we use? We can cross that bridge when we come to it.

**Comment by snarfed** [comment]
Hmm, subdomains right now are used to identify accounts, ie identity/protocol, not types of records. Eg we currently handle both `Article`s and `Note`s, and we don't split those out onto separate subdomains in either web or ActivityPub.

What use case were you hoping that subdomain-per-lexicon for ATProto would provide for?

**Comment by Tamschi** [comment]
Ah no, not per lexicon, but per app. For example, I may want to follow someone's blog posts but not their microblog posts.

On ActivityPub this is easy, since that person will generally have distinct e.g. Mastodon and Ghost[^1] accounts and I can follow just one. (The distinction is somewhat enforced by the ActivityPub C2S protocol not being full-featured enough afaik, so apps are vertically somewhat integrated in a way that covers identity.)

On ATProto this is also easy, since while the expectation is that the DID is shared between services, apps generally display only their relevant content. WhiteWind specifically doesn't seem to have a concept of following someone though, it lets you cross-post an announcement as Bluesky post instead.

However, cross-network there's an issue if WhiteWind content is bridged just on `bsky.`, since then I won't be able to follow only their blog posts using Mastodon and most other fediverse software, since they filter both incoming `Note`s and `Article`s into the same local representation and don't offer a filter in that regard.

It would be better if I saw someone's Bluesky and WhiteWind activity on separate ActivityPub-side bridge accounts and could navigate between them using a mention in the respective profiles. Since WhiteWind seems to have no notion of following or blocking someone and no Chat support, I think it would be necessary to configure the bridge regarding WhiteWind activity using the Bluesky Chat interface though.

[^1]: Apparently Ghost is getting some microblogging features, but I image most will keep that to blog announcements.

**Comment by snarfed** [comment]
You're right that people in the fediverse generally separate different content types onto different accounts, which makes it easy to control and bridge them separately. Personally, I think that's kind of a weak apology for a somewhat broken ecosystem - it's not that I _choose_ to post blog posts, or events, etc on a separate account than my Mastodon, it's that I literally _can't_, https://snarfed.org/?s=fediverse+functionality - but it is the fedi status quo, for better or worse.

ATProto and many other more modern decentralized social protocols are different: they tend to decouple content types and app semantics from user data storage. Eg in ATProto, user data is in PDSes but semantics and data types are handled by appviews and clients, in Nostr and Farcaster semantics are pretty much all in clients, etc. (Personally, I think this is better for everyone. You can still use multiple accounts and alts if you want, and you're not limited by content type! But that's just my opinion.)

A complicating factor in these more-modern protocols is that there isn't necessarily a one-to-one mapping between content types, eg ATProto lexicons, and apps. A number of ATProto apps share lexicons, either each other's (eg `app.bsky.*`) or lexicons like https://lexicon.community/ that are deliberately intended to be shared and cross-app.

When Bridgy Fed sees a new record in a user's repo, it won't see which client wrote it, and it can't know which appview "owns" it. There sometimes won't even be a clear "owner" app.

So, I'm all for giving users more control over how, when, and which of their stuff is bridged where! Content type could definitely be one way. But outside of the fediverse, I'm not really able to do it by "app" or "client."

**Comment by Tamschi** [comment]
By "app" I mean roughly lexicon namespaces, like `app.bsky.…`, not the implementation that writes the record. (It's awfully hard to be precise about that, but generally I will say "AppView" or "client"/"frontend" when talking about the latter.)

Follows are app-specific in ATProto too, rather than being global between identities, so I think app-granularity maps a lot better to ActivityPub Actors than one-Actor-per-DID does even just in terms of equivalence and less lossy conversion.

(There is a case for merging them where one ATProto app uses another's follow-records specifically, but I have not encountered that yet.)

**Comment by Tamschi** [comment]
As a side-note: I don't think the ActivityPub model is broken, just different. It to me has less friction in terms of fluid, contextual identities.

(That may just be a UX difference, first and foremost, but I think there's some design priorities that shine through in Bluesky burying the account switcher in Settings.)

**Comment by manton** [comment]
I’m exploring this for Micro.blog too and wrote this post:

https://www.manton.org/2025/07/02/next-step-with-at-proto.html

I’ve also heard that WhiteWind is no longer maintained. Leafpub seems to be closest to long-form blogging.

**Comment by snarfed** [comment]
Also now https://standard.site/

**Comment by snarfed** [comment]
Did some homework on where we are internally. I think we're ready to convert and bridge one object/activity to multiple ATProto records. May have some quirks around the edges, but in general everything should still Just Work.

As for which of the many lexicons we choose to bridge to, we don't have explicit rules or policy around that yet, just product priority  judgment calls for now. First up is probably standard.site, #2324.

So, let a thousand flowers bloom! I'll close this meta-issue, and we can start filing and implementing feature request issues for individual lexicons as they come up.

(Letting people choose which of a user's post types to follow is a valid feature request! ...but as usual, we try to follow networks' native functionality in Bridgy Fed, not invent or lead it, so we'd look to see networks provide this natively first, ideally in-protocol, and follow that.)
